Extend cable cell probe variety.

halfflat opened the pull request:
Fixes #589 and #730, providing cell-wide probes and a correct total trans-membrane current.
Fixes #822, providing spatial interpolation of membrane voltages as governed by cable resistivity and diameter.

* Add partition of point mechanism target vector by cell index to `fvm_mechanism_config`.
* Fix sign error in axial current interpolation.
* Replace `fvm_probe_info` with class that wraps a variant type capturing one of a set of backend-cellgroup probe translation representations: `fvm_probe_scalar`, `fvm_probe_interpolated`, `fvm_probe_multi`, `fvm_probe_weighted_multi`, `fvm_probe_membrance_currents`.
* Refactor `fvm_lowered_cell_impl::resolve_probe_address` so that it avoids the long chain of if-else tests on wrapped `any` type with an invocation of `util::any_visitor`. Each specific cable cell probe address representation is translated into one of the intermediate `fvm_probe_info` variants via an overload of `resolve_probe`.
* Add new non-scalar cell probe address types for all-of-cell values, including point mechanism state, density mechanism state, ion currents and concentrations, membrane voltage by CV cable, and total trans-mebrane currents. The latter are ultimately calculated by computing the net axial current fluxes in each CV.
* Add `util::any_ptr` field to the sampling function interface, pointing to constant-per-simulation-run metadata associated with a probe address. Underlying metadata type varies between probe types.
* Extend `simple_sampler` so that it optionally captures sampler metadata.
* Add special case to `simple_sampler` for vector-valued sample data, allowing it to translate from a pointer-pair range to a std::vector.
* Fix minor bugs in `util::variant`.
* Add to `mc_cell_group` the logic for translating each intermediate probe representation to a sampler callback invocation, reusing temporary scratch vectors across calls.
* Add unit comments to `embed_pwlin_data`.
* Add probe variety demonstration `probe-demo` to examples, covering most of the new cable cell probe classes.
* Add `util::any_visitor` runtime-type dependent dispatcher class.
* Add `util::overload` helper for generating overloaded functional objects from individual functions, lambdas or functionals.
* Extend probe unit tests in `test_probe.cpp` to cover new probe address types.
* Add unit tests for `any_visitor` and `overload`.

halfflat commented:
I'll make documentation edits in a separate PR, once this PR has been resolved, in case changes to the API are required.

bcumming commented on arbor/include/arbor/cable_cell.hpp:
```diff
@@ -29,51 +29,153 @@ struct lid_range {
+// Ionic internal concentration [mmol/L] of ion `source` across components of the cell.
+// Sample value type: `cell_sample_range`
+// Sample metadata type: `mcable_list`
+struct cell_probe_ion_int_concentration_cell {
```
These `cell_X_cell` types are a little difficult to parse.

`cable_X_cell` and `cable_X` to indicate that we are talking about a cable cell specifically?

halfflat replied:
Happy to rename `cell_probe_FOO` to `cable_probe_FOO`.

bcumming commented on arbor/include/arbor/cable_cell.hpp:
```diff
@@ -29,51 +29,153 @@ struct lid_range {
+// Total ionic current [nA] attributed to the ion `source` across components of the cell.
+// Sample value type: `cell_sample_range`
+// Sample metadata type: `mcable_list`
+struct cell_probe_ion_current_cell {
```
The explosion of types makes the interface unwieldy (according to my personal tastes).

The challenge is to make documentation and an interface that users can quickly get to what they want (particularly the python wrapper) without spending much time having to read docs and look for previous examples.

I think a "stringy" interface might fit the bill. Do you think it should be implemented in C++, on top of these types, or only in the Python wrapper (I would still write the string->probe-type code in C++, but keep it in `arbor/python` to be called directly by python wrapper calls).

halfflat replied:
For the C++ API, it's this, or an enum together with a bunch of enum-contigent variant values. The type interface is, I am certain, clearer and easier to follow than the alternative.

What we do in Python though is another story.

bcumming commented on arbor/mc_cell_group.cpp:
```diff
@@ -79,6 +80,245 @@ void mc_cell_group::set_binning_policy(binning_kind policy, time_type bin_interv
+}
+
+void reserve_scratch(fvm_probe_scratch& scratch, std::size_t n) {
+    tuple_foreach([n](auto& v) { v.reserve(n); }, scratch);
```
Given that `tuple_foreach` is used only once, right below where it is defined, I think it would be easier to just say:
```C++
scratch.get<0>().reserve(n);
scratch.get<1>().reserve(n);
```

Looking at `fvm_probe_scratch` and how it is used, it looks like it could be a simple struct. Is there reason it is a `tuple`?

halfflat replied:
At one point, there were more vectors in scratch, and a struct was getting unwieldy and copy/pastey.

With just the two, a struct is not bad. But the tuple approach has the benefit of being easy to generalize. It doesn't strike me as being particularly obscure as a tuple, but that's probably a matter of taste.

bcumming commented on example/probe-demo/probe-demo.cpp:
```diff
@@ -0,0 +1,247 @@
+
+        cable_recipe R(opt.probe_addr, opt.n_cv);
+
+        auto context = arb::make_context();
```
Do we explicitly all the new probe features on GPU?

Can we add a flag to "turn on" running the sim on the GPU if one is available?

halfflat replied:
I'm not sure I follow what you'd prefer here ... as things stand, the CPU ones are always tested, and the GPU versions are tested if there is one available and we've built GPU support in.

halfflat replied:
Regarding the probe example, this comes back to the ongoing debate about the easy context stuff, #988.

bcumming commented on test/unit/test_probe.cpp:
```diff
@@ -472,28 +1099,59 @@ TEST(probe, gpu_v_i) {
+TEST(probe, gpu_v_cell) {
+    context ctx = make_context(proc_allocation{1, arbenv::default_gpu()});
+    if (has_gpu(ctx)) {
+        run_v_cell_probe_test<gpu::backend>(ctx);
```
Well, that answers my question about whether the GPU backend was being tested.

halfflat replied:
Ah, yes :)

halfflat commented on arbor/fvm_layout.cpp:
```diff
@@ -433,6 +436,12 @@ fvm_cv_discretization fvm_cv_discretize(const cable_cell& cell, const cable_cell
             D.init_membrane_potential[i] /= D.cv_area[i];
             D.temperature_K[i] /= D.cv_area[i];
         }
+        else if (p!=-1) {
```
This will only turn up if there is a trivial CV at the root with multiple top level branches. There isn't really a consistent choice of how to pick a starting membrane potential or temperature here, though some heuristic could be picked. Pick the first branch values, perhaps?

halfflat commented on arbor/fvm_lowered_cell_impl.hpp:
```diff
@@ -648,111 +653,367 @@ fvm_size_type fvm_lowered_cell_impl<Backend>::fvm_intdom(
+fvm_probe_info resolve_probe(const cell_probe_total_current_cell& p, probe_resolution_data<B>& R) {
+    fvm_probe_membrane_currents r;
+
+    auto cell_cv_ival = R.D.geometry.cell_cv_interval(R.cell_idx);
```
`D` for the `fvm_discretiazation` is used consistently through `fvm_layout.cpp` and the rest of `fvm_lowered_cell_impl.hpp`. Using a longer name makes some of these expressions even more unwieldy than they are, and I'm pretty sure anyone with their hands in the code will quickly see how `D` is used.

We can make it longer, but I don't think it will consequently aid readability.

halfflat commented on arbor/fvm_lowered_cell_impl.hpp:
```diff
@@ -648,111 +653,367 @@ fvm_size_type fvm_lowered_cell_impl<Backend>::fvm_intdom(
+    typename Backend::shared_state* state;
+    const cable_cell& cell;
+    const std::size_t cell_idx;
+    const fvm_cv_discretization D;
```
This should be a const ref.

halfflat commented:
Updated PR addressing review (see commit comments on 57d8bcce2e27d5bc1ae59cddf52dd76b5e8fbd5f), plus added documentation for the sampling API changes and new probe types.
